Multi class and Multi label segmentation

TimPaquaij opened the pull request:
# What does this PR do ?
This PR makes it possible to use Multi-Class Segmentation and Multi-label Segmentation. Furthermore, several losses have been added for unbalanced segmentation tasks.

**Collections**:
MTL
# Changelog
- There are too many changes to add specific lines

# Usage
* See train/mtrls.yaml for usage within the model
* see tests/segmentation/losses for using the separate losses

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/wdika/atommic/blob/main/CONTRIBUTING.md)

- [x] Did you write any new necessary tests?
- [] Did you add or update any necessary documentation?
- [] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [] Reviewer: Does the PR have correct import guards for all optional libraries?

**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.


## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/wdika/atommic/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

wdika reviewed (changes requested):
@TimPaquaij please check my comments and changes.

Mind removing extra lines and spaces, while maintaining the style with spaces and lines. For example:

- variables should be set like "x = 1" but not "x= 1" or "x =1".
- Between classes and functions definitions leave 2 empty lines.
- On new function definition inside a class leave 1 empty line.

Also,
- [ ] Add __init__.py on tests/collections/segmentation/losses

TimPaquaij commented on atommic/collections/multitask/rs/nn/base.py:
```diff
@@ -37,8 +37,9 @@
 from atommic.collections.reconstruction.metrics import mse, nmse, psnr, ssim
-from atommic.collections.segmentation.losses.cross_entropy import CrossEntropyLoss
-from atommic.collections.segmentation.losses.dice import Dice
+from atommic.collections.segmentation.losses.cross_entropy import CategoricalCrossEntropyLoss, BinaryCrossEntropyLoss
```
I will fix this

TimPaquaij commented on requirements/requirements.txt:
Previously no version was specified, so I think it automatically downloaded a version > 0.20.3. I tested 0.20.3 and this is working for me as well. However, newer versions have renamed some functions which will lead to errors.

wdika replied:
Okay, I think it's good to stick to version huggingface_hub==0.20.3 then

wdika commented on atommic/collections/common/data/mri_loader.py:
What's the difference between `len(self.examples)` and `[example[1] for example in self.examples]`?

wdika commented on atommic/collections/multitask/rs/nn/base.py:
@TimPaquaij now `self.segmentation_type = cfg_dict.get("segmentation_type", "MLS")` why do we need to set it to mls? This the the multitask collection for rs (reconstruction and segmentation).

wdika commented on atommic/collections/multitask/rs/nn/base.py:
```diff
@@ -155,8 +157,10 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):  # noqa: MC0001
         self.log_multiple_modalities = cfg_dict.get("log_multiple_modalities", False)
 
         # Set threshold for segmentation classes. If None, no thresholding is applied.
+        self.segmentation_type = cfg_dict.get("segmentation_type", "MLS")
```
@TimPaquaij now self.segmentation_type = cfg_dict.get("segmentation_type", "MLS") why do we need to set it to mls? This the the multitask collection for rs (reconstruction and segmentation).

wdika replied:
@TimPaquaij maybe it's obvious, but now self.segmentation_type = cfg_dict.get("segmentation_type", "MLS") why do we need to set it to mls by default?

wdika commented on tests/collections/multitask/rs/models/test_mtlrs.py:
@TimPaquaij shouldn't we have a separate ``test_mtlms.py`` file?

wdika reviewed (changes requested):
@TimPaquaij apart from minor changes I am thinking if we should have a `multitask/mls` **collection** instead of merging it to `multitask/rs`. What do you think?

wdika commented:
@TimPaquaij please also check the CodeCov/Tox errors. It seems like there is an issue with importing the new losses.

TimPaquaij commented:
I think it is not valid to make a new collection, since that would suggest that this is a different task. Whereas I think it is a different approach to achieving the same task. For some datasets, the multi-label approach would be more beneficial, think of detecting strokes within the brain. In this case, you have 2 labels which can be both valid. One pixel can be brain and there could be a lesion. However, even for this dataset, you can also determine the areas of the lesions with a multi-class approach since eventually, we are interested in the locations and the size of the specific lesions.

wdika commented on atommic/collections/common/data/mri_loader.py:
@TimPaquaij what's the difference between len(self.examples) and [example[1] for example in self.examples]?

wdika commented on atommic/collections/segmentation/nn/base.py:
```diff
@@ -77,6 +79,8 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
         # Set threshold for segmentation classes. If None, no thresholding is applied.
         self.segmentation_classes_thresholds = cfg_dict.get("segmentation_classes_thresholds", None)
         self.segmentation_activation = cfg_dict.get("segmentation_activation", None)
+        self.segmentation_type = cfg_dict.get("segmentation_type", "MLS")
```
@TimPaquaij same here as in rs. Maybe it's obvious, but now self.segmentation_type = cfg_dict.get("segmentation_type", "MLS") why do we need to set it to mls by default?

wdika commented on projects/MTL/rs/SKMTEA/conf/train/mtlrs.yaml:
```diff
@@ -49,6 +49,8 @@ model:
   reconstruction_module_keep_prediction: true
   reconstruction_module_accumulate_predictions: true
   segmentation_module: AttentionUNet
+  segmentation_type: MLS
```
@TimPaquaij  is this always the default now?

TimPaquaij commented:
So to be clear, the original segmentation method you implemented was based on a multi-label segmentation (MLS) approach. Where, each class is handled separately and is not influenced by another class (see thresholding in inference_step). For the SKMTEA dataset and for the upcoming spatially adaptive semantic guidance module (SASG), the segmentation needs to be in a multi-class segmentation form where the classes are 'related' (and can be transformed into probabilities). So this is why I wanted to have a variable which can be set to a specific segmentation approach. Since then I can use this variable later on to create warnings when different techniques are combined which can not be combined.

wdika commented:
> So to be clear, the original segmentation method you implemented was based on a multi-label segmentation (MLS) approach. Where, each class is handled separately and is not influenced by another class (see thresholding in inference_step). For the SKMTEA dataset and for the upcoming spatially adaptive semantic guidance module (SASG), the segmentation needs to be in a multi-class segmentation form where the classes are 'related' (and can be transformed into probabilities). So this is why I wanted to have a variable which can be set to a specific segmentation approach. Since then I can use this variable later on to create warnings when different techniques are combined which can not be combined.

This is also different from combining the classes like in the [process_segmentation_labels](https://github.com/wdika/atommic/blob/d34fcad4d577a4a373c8fd80d757121bd65d53af/atommic/collections/segmentation/data/mri_segmentation_loader.py#L172), right?

TimPaquaij commented:
So multi-label and multi-class is a type of approach. So it includes the way you should pre-process your data, determine the loss and asses the final prediction. In the case of multi-class segmentation it is not possible that an pixel is defined as no class. So we need to make sure that the background/not a class is also defined as an seperate class. Whereas the multi-label does not need this and can have the same number of classes

wdika commented:
> So multi-label and multi-class is a type of approach. So it includes the way you should pre-process your data, determine the loss and asses the final prediction. In the case of multi-class segmentation it is not possible that an pixel is defined as no class. So we need to make sure that the background/not a class is also defined as an seperate class. Whereas the multi-label does not need this and can have the same number of classes

FYI when I start training for MLS/MCS I get the warning about the weights. Maybe it has to do with the yaml file you shared with me but please confirm this is intended and doesn't break anything for MLS/MCS.

> [atommic W atommic_logging:336] atommic/atommic/collections/multitask/rs/nn/base.py:184: UserWarning: Sum of segmentation losses weights is not 1.0. Adjusting weights to sum up to 1.0.
      warnings.warn("Sum of segmentation losses weights is not 1.0. Adjusting weights to sum up to 1.0.")

wdika commented on atommic/collections/segmentation/nn/base.py:
```diff
@@ -685,6 +740,14 @@ def test_step(self, batch: Dict[float, torch.Tensor], batch_idx: int):
+            if self.segmentation_output_mode == "binary":
+                predictions = predictions.argmax(dim=1, keepdim=True)
+                predictions = one_hot(predictions, num_classes=self.segmentation_module_output_channels, dim=1)
+            else:
```
@TimPaquaij the else doesn't make sense here. What's the logic?

wdika replied:
I think the intent is wrong here

wdika commented on atommic/collections/multitask/rs/nn/base.py:
```diff
@@ -1475,6 +1556,17 @@ def test_step(self, batch: Dict[float, torch.Tensor], batch_idx: int):
+                predictions_segmentation = one_hot(
+                    predictions_segmentation, num_classes=self.segmentation_module_output_channels, dim=1
+                )
+            else:
```
@TimPaquaij the else doesn't make sense here. What's the logic?

wdika replied:
I think the intent is wrong here

wdika commented:
@TimPaquaij I tried to implement MLS/MCS for single-task segmentation models and I found when computing the loss in [process_segmentation_loss](https://github.com/wdika/atommic/blob/ce34aefd86a7c307a21fb6cdf8bf41d04617c7c6/atommic/collections/segmentation/nn/base.py#L328) it will return a tensor of shape == len(classes), meaning the loss of each class. So then it will fail to get the loss.item() because it doesn't contain only one value.

Then I checked back in the MTLRS collection the same [function](https://github.com/wdika/atommic/blob/ce34aefd86a7c307a21fb6cdf8bf41d04617c7c6/atommic/collections/multitask/rs/nn/base.py#L630) and I saw that it will first return the loss tensor of shape == len(classes) and then it will select only the first class loss due to the hardcoded fix `loss = loss[1][0]` for the base dice loss we had before. Where loss[1] is the dice loss and not the dice metric (which is the loss[0]) and loss[1][0] is just the loss of the first class.

Could it be that now this is not implemented in the right way?

TimPaquaij commented:
I think this is the result of me changing the batch size argument within the dice. In de docstring it says use this argument to get batch_size, whereas in de docstring of the dice loss on monai is was inverted. So what I did is inverting the logic by adding 'not' infront of each statement to get a correct logic with the docstring. When training place make sure to double check the parameters of the dice loss. Before you had batch_size on true and reduction on mean_batch, which resulted in a singels loss value. Since I changed the batch argument this now results in a loss with len(classes). So make sure to change the reduction to mean (reduction of both batch and channel). Or change the batch_size to false and the reduction to mean_channel. Please view test_dice to see the possibilities.

Why I changed this into loss[1][0] is because after a valid reduction the single loss of the dice is still in a list, which is fine when using only one loss but for combining it with other segmentation losses you will get an error

wdika commented:
> This resulted from changing the batch size argument within the dice. The docstring says to use this argument to get batch_size, whereas, in the docstring of the dice loss on monai, it was inverted. So, I did invert the logic by adding 'not' in front of each if statement to get the correct logic corresponding to the docstring. When training, please make sure to double-check the parameters of the dice loss before you had batch_size on true (resulted in the reduction of the batch_size) and reduction on mean_batch (which resulted in the reduction of the channel size since this was the first dimension after the batch_size reduction), which resulted in a single loss value. Since I changed the batch argument, this has now resulted in a loss with len(classes). So make sure to change the reduction to mean (reduction of both batch and channel). Or change the batch_size to false and the reduction to mean_channel. Please take a look at test_dice to see the possibilities.
> 
> Why I changed loss[1] into loss[1][0] because, after a valid reduction, the single loss of the dice is still in a list, which is fine when using only one loss, but when combining it with other segmentation losses, you will get an error

So `dice_loss_reduction` and `dice_metric_reduction` should be always set to `mean` instead of `mean_batch` from now on? Also are we sure the `loss[1][0]` doesn't break previous and single-task segmentation?

TimPaquaij commented:
Yes indeed, all the reductions need to be set to 'mean' instead of 'mean_batch' when batch_size is set to 'true'!

wdika commented:
> Yes indeed, all the reductions need to be set to 'mean' instead of 'mean_batch' when batch_size is set to 'true'!

You mean `dice_loss_batch` and `dice_metric_batch` right? We really need to right down all those things upon merging the PR.

wdika reviewed (commented):
- [ ] Validate previous behavior on segmentation loss is maintained.
- [ ] Write detailed notes for the implementation.
- [x] Create new yaml files.
- [x] Update docs.

TimPaquaij commented:
Okay I changed loss[1][0] into loss[1].squeeze() This makes sure if there is a batch or a channel dimension it does not interfere.This only makes sure when the loss is equal to a 0 dimension loss the loss --> torch.tensor([0.123]) will be transformed into torch.tensor(0.123) needed to combine with other loss. I think this is a better implementation
